### PR TITLE
Data class for parsing JEDI log file. Adds capability to extract convergence info

### PR DESCRIPTION
**Pull request by danholdaway**
## Description

Add a new data class for parsing the JEDI log files. It should be fairly robust and able to handle most of the minimizer logs that JEDI outputs. It will make sure not to pick up GMRES output. 

Later on the code can be extended to parse other things from the log files, such as obs statistics, timing etc.

- Added new LinePlot class to Eva.
- Improved the error reporting for when the base class imports with error.
- Removed `:` from the test file names. Use of `:` comes from R2D2 and upsets some filesystems.

The new added test produces the following plots:

### DRIPCG logs from ctest. Log includes GMRES, which needs to be ignored.

norm_gradient_reduction_dripcg_ctest

cost_function_dripcg_ctest

### RPCG log non ctest and no output of cost function

norm_gradient_reduction_rpcg

### Example of the variables added to the data collection:

```
INFO DataCollections: Collections available: 
INFO DataCollections: 
INFO DataCollections: Collection name: jedi_log_test
INFO DataCollections: 
INFO DataCollections:  Dimensions:
INFO DataCollections:   dim_0: 15
INFO DataCollections: 
INFO DataCollections:  Coordinates:
INFO DataCollections: 
INFO DataCollections:  Data (group::variable):
INFO DataCollections:   convergence::total_iteration               (int32  ) | Min=         +1, Max=        +15
INFO DataCollections:   convergence::inner_iteration               (int32  ) | Min=         +1, Max=        +10
INFO DataCollections:   convergence::gradient_reduction            (float32) | Min=+1.7911e+04, Max=+1.0963e+05, RMS=+5.1458e+04
INFO DataCollections:   convergence::norm_reduction                (float32) | Min=+1.2612e-01, Max=+7.7191e-01, RMS=+3.9423e-01
INFO DataCollections:   convergence::j                             (float32) | Min=+5.9964e+03, Max=+6.3174e+03, RMS=+6.0667e+03
INFO DataCollections:   convergence::jb                            (float32) | Min=+2.5703e-05, Max=+1.1133e-03, RMS=+7.8399e-04
INFO DataCollections:   convergence::jojc                          (float32) | Min=+5.9964e+03, Max=+6.3174e+03, RMS=+6.0667e+03
INFO DataCollections:   convergence::outer_iteration               (int32  ) | Min=         +1, Max=         +2
INFO DataCollections:   convergence::gradient_reduction_normalized (float32) | Min=+1.6338e-01, Max=+1.0000e+00, RMS=+4.6939e-01
INFO DataCollections:   convergence::norm_reduction_normalized     (float32) | Min=+1.6338e-01, Max=+1.0000e+00, RMS=+5.1072e-01
INFO DataCollections:   convergence::j_normalized                  (float32) | Min=+9.4919e-01, Max=+1.0000e+00, RMS=+9.6032e-01
INFO DataCollections:   convergence::jb_normalized                 (float32) | Min=+2.3087e-02, Max=+1.0000e+00, RMS=+7.0419e-01
INFO DataCollections:   convergence::jojc_normalized               (float32) | Min=+9.4919e-01, Max=+1.0000e+00, RMS=+9.6032e-01
```

### CODE REVIEW

File: src/eva/data/jedi_log.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        with open(jedi_log_to_parse) as jedi_log_to_parse_open:
+            jedi_log_text = jedi_log_to_parse_open.read()
+
+        # Split log into list of lines
+        jedi_log_lines = jedi_log_text.split('\n')
+
+        # Check if this was a ctest and if so determine test prepend string
+        test_string = ''
+        for jedi_log_line in jedi_log_lines:
+            if jedi_log_line[0:4] == 'test':
```
Comment (CoryMartin-NOAA) [inline src/eva/data/jedi_log.py:61]: I thought it was `Test        :` like `CostJo    :` ? or am I thinking of something else?

File: src/eva/tests/config/testJediLog.yaml [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            add_ylabel: 'Gradient reduction'
+            layers:
+            - type: LinePlot
+              x:
+                variable: jedi_log_test::convergence::total_iteration
+              y:
+                variable: jedi_log_test::convergence::gradient_reduction
+              color: 'black'
+
+          - add_xlabel: 'Total innter iteration number'
```
Comment (CoryMartin-NOAA) [inline src/eva/tests/config/testJediLog.yaml]: typo

File: src/eva/tests/config/testJediLog.yaml [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            add_ylabel: 'Norm reduction'
+            layers:
+            - type: LinePlot
+              x:
+                variable: jedi_log_test::convergence::total_iteration
+              y:
+                variable: jedi_log_test::convergence::norm_reduction
+              color: 'black'
+
+          - add_xlabel: 'Total innter iteration number'
```
Comment (CoryMartin-NOAA) [inline src/eva/tests/config/testJediLog.yaml]: typo

File: src/eva/tests/config/testJediLog.yaml [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  graphics:
+    - figure:
+        layout: [3,1]
+        figure size: [12,10]
+        title: 'Gradient and Norm Reduction Plots DRIPCG'
+        output name: jedi_log/convergence/norm_gradient_reduction_dripcg_ctest.png
+      plots:
+
+
+        - add_xlabel: 'Total innter iteration number'
```
Comment (CoryMartin-NOAA) [inline src/eva/tests/config/testJediLog.yaml]: typo

File: src/eva/tests/config/testJediLog.yaml [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+          layers:
+          - type: LinePlot
+            x:
+              variable: jedi_log_test::convergence::total_iteration
+            y:
+              variable: jedi_log_test::convergence::gradient_reduction
+            color: 'black'
+
+
+        - add_xlabel: 'Total innter iteration number'
```
Comment (CoryMartin-NOAA) [inline src/eva/tests/config/testJediLog.yaml]: typo

File: src/eva/tests/config/testJediLog.yaml [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+          layers:
+          - type: LinePlot
+            x:
+              variable: jedi_log_test::convergence::total_iteration
+            y:
+              variable: jedi_log_test::convergence::norm_reduction
+            color: 'black'
+
+
+        - add_xlabel: 'Total innter iteration number'
```
Comment (CoryMartin-NOAA) [inline src/eva/tests/config/testJediLog.yaml]: typo

File: src/eva/tests/config/testJediLog.yaml [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+            label: 'Normalized norm reduction'
+
+    - figure:
+        layout: [3,1]
+        figure size: [12,10]
+        title: 'Cost Function Plots DRIPCG'
+        output name: jedi_log/convergence/cost_function_dripcg_ctest.png
+      plots:
+
+        - add_xlabel: 'Total innter iteration number'
```
Comment (CoryMartin-NOAA) [inline src/eva/tests/config/testJediLog.yaml]: I'll stop pointing it out now, find -> replace `innter` to `inner`.

**Review by CoryMartin-NOAA — COMMENTED** [review]
A few minor comments, but thanks a lot for adding this!

File: src/eva/tests/config/testJediLog.yaml (hunk 2 above)
Comment (danholdaway) [reply src/eva/tests/config/testJediLog.yaml]: Done. Thanks for spotting that.

File: src/eva/data/jedi_log.py (hunk 1 above)
Comment (danholdaway) [reply src/eva/data/jedi_log.py:61]: From what I can see when running ctest is prepends the log with a few lines. One of the first lines looks like e.g.:

`test 80`

We need to get `80` in order to know that all the other lines of the log will have `80: ` prepended to them. The next step is to remove that from all the lines so the subsequent parsing can proceed similarly to if the log came from a non ctest run, which would presumably be the standard.

Comment (danholdaway) [reply src/eva/data/jedi_log.py:61]: I believe the `Test        :` bit might be from when you output things with `oops::Log::test()`? As in [this line](https://github.com/JCSDA-internal/oops/blob/66c008ca38ae342fb8fda9b283cd16bb8c74c64c/src/oops/runs/ConvertState.h#L133). Then you get a line which looks like `80: Test     :`.

Comment (CoryMartin-NOAA) [reply src/eva/data/jedi_log.py:61]: ah okay, got it

**Review by CoryMartin-NOAA — APPROVED** [review]

**Pull request merged**